server/protocols/webadmin/webadmin-jmap/src/main/java/org/apache/james/webadmin/data/jmap/MessageFastViewProjectionCorrector.java

```diff
@@ -144,6 +153,22 @@ boolean failed() {
             .then();
     }
 
+    Mono<Void> correctMessageProjectionItems(MessageId messageId) {
+        try {
+            MailboxSession session = mailboxManager.createSystemSession(FAST_VIEW_CORRECTOR_USER);
+            return Flux.fromStream(mailboxSessionMapperFactory.getMessageIdMapper(session)
```

Comment:
TL;DR
you should never use a `Mapper` outside mailbox components.

Long explanation:
The abstraction to access a Mailbox is what's in `mailbox-api` (MailboxManager and others `Manager`s)
For some (historical) reasons, all mailboxes implementations rely on some abstract classes located in `mailbox-store` (`Mapper`s are part of it).
It allows, to some extent, to share some business logic between mailbox implementations.
The design is the following: a MailboxManager implementation has to extend StoreMailboxManager which requires the implementation of some `Mapper`s.

![PlantUML model](http://www.plantuml.com/plantuml/png/TP4n2iCm34Ltdq8No0LAA87EcNe2nLQBmRLTMIaTUlWACZd8YmV_zvy4MfeWo_vABcOXVg0dM36chNnNpFWaThWrOVG2Fc5hS9V2T4AEvAOFvi3OfvQvLgFLXQwNmFoR9dEsyNV0etMwEIDKCrkjT2-um1MOtdjaQiuMmNWRhSzGkNnYqFf2EUWj_W00)

There are (at least) two problems with this design:

* MailboxManager is an interface implemented exactly once, it means it should/could disappear to reduce useless complexity (and at the same time it's the abstraction that we want to favor because it's the API with design with care and that actually match the business use cases)
* MailboxMapper / StoreMailboxManager is a poor implementation solution because MailboxMapper is designed to provide low-level functions with some implicit requirements like "we can do transactions or locks". In some cases, there are no solution at the MailboxMapper level to solve a design issue we these requirements are not met.

The right design for this is:

* favor MailboxManager and `mailbox-api` interfaces (and never use anything from `mailbox-store` module outside a mailbox implementation)
* ensure correctness of `Manager`s implementation by providing contract test and not by sharing abstract classes (composition is obviously ok)

Reply:
We do agree with that, there's just an issue in `MailboxManager` API : we cannot act on a single message without knowing a user it belongs to.

To be relying on mailbox-api, we need to : 
  - grant access to system sessions, and trust them.
  - Review system session usages to ensure we do not introduce vulnerability.

Similar issue can then be takkled in reIndexing as well.

Sounds like the good time to takkle this technical dept reduction concerns.

Reply:
I know we talked about this with @chibenwa yesterday, but I forgot to mention it in the PR, so thanks for raising it @mbaechler .

Yes that would deserve a separate refactoring task

Reply:
So what is the plan exactly?

Reply:
I started working on a PR with @chibenwa 's suggestion above.

Reply:
cf #3038